### PR TITLE
Add getservbyname and getservbyport

### PR DESCRIPTION
**Pull request by ahgamut**
`getservbyname` and  `getservbyport` have implementation similar to that of `gethostbyname` and `gethostbyaddr`. 
output matches glibc, except that the aliases for a service (ie `struct servent.s_aliases`) are not filled.

There is no `struct ServicesTxt` like `struct HostsTxt`. For every call of `getservbyname`/`getservbyport` the lookup of `/etc/services` is done by opening the file and parsing each line one-by-one to compare. This is slow, but the implementation is simple; the file loading is similar to `GetHostsTxt` and the file parsing is similar to `ParseHostsTxt`.

The remaining functions are stubs in the [`musl`](https://github.com/ifduyue/musl/blob/master/src/network/serv.c) source code, so I left them unimplemented here as well.

### CODE REVIEW

**Comment by ahgamut** [comment]
@jart would you prefer a `struct HostsTxt`-style implementation? I skipped that because I saw `TODO`s for some of the internal stuff used for setting up `struct HostsTxt` (for eg. `concat`/`append` vs `CONCAT`/`APPEND`).

**Comment by ahgamut** [comment]
What is the maximum length of a service name in `/etc/services`?
[RFC 6335](https://datatracker.ietf.org/doc/html/rfc6335#section-5.1) says it is 15 characters, but [this](https://www.ibm.com/docs/en/zos/2.1.0?topic=assignments-etcservices-etcservices-port) says it is 32. 

If the bound is known, we can avoid  `malloc` calls in the `getservby*` functions as well.

**Review by jart — APPROVED** [review]
Thanks for sending this! Looks good to merge.

I noticed there's no unit tests for this change. Here's how that could have worked:

```c
#include "libc/testlib/testlib.h"

char testlib_enable_tmp_setup_teardown;

void SetUp(void) {
  int fd;
  const char *s = "\
# hello\n\
finger          79/tcp\n\
http            80/tcp          www             # WorldWideWeb HTTP\n";
  ASSERT_NE(-1, (fd = creat("hosts", 0755)));
  ASSERT_NE(-1, write(fd, s, strlen(s)));
  ASSERT_NE(-1, close(fd));
}

TEST(ftruncate, test) {
  ASSERT_EQ(80, LookupServicesByName("http", ptr, "hosts"));
}
```

**Comment by jart** [comment]
> @jart would you prefer a struct HostsTxt-style implementation? I skipped that because I saw TODOs for some of the internal stuff used for setting up struct HostsTxt (for eg. concat/append vs CONCAT/APPEND).

This is fine for now. We can revise it to have a global singleton as soon as it presents performance issues.

> What is the maximum length of a service name in /etc/services?

Our malloc() is very fast but this does look like a case where it's not needed. If RFC6335 says it's 15 characters then it must be 15 characters. I'm fine enforcing that limit and having a `char[16]` buffer. But, due to how people are:

image

I would support your decision if you chose to make the buffer larger. It doesn't change the fact that the number is officially still 15. It's more of a Postel's Maxim kind of thing in being generous towards nonconformity. But in that case I'd argue that unlimited malloc is much better than defining another arbitrary but somewhat higher limit.

File: libc/dns/servicestxt.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+ * format of /etc/services is like this:
+ *
+ *      # comment
+ *      # NAME      PORT/PROTOCOL   ALIASES
+ *
+ *      ftp		    21/tcp
+ *      fsp		    21/udp		    fspd
+ *      ssh		    22/tcp
+ *
+ * @param servport is the port number (in network byte order)
```
[the remark this answers is not in the document]
Comment (ahgamut) [reply libc/dns/servicestxt.c:66]: I didn't realize it was network byte order until a simple test for `getservbyport` with `port = 22` failed on glibc and musl; had to re-read the [`man`](https://linux.die.net/man/3/getservbyport) page entry for `getservbyport` to be sure. Nice trivia, though.

File: libc/dns/servicestxt.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * Opens and searches /etc/services to find port for a given name.
+ *
+ * @param servname is a NULL-terminated string
+ * @param servproto is a pointer to a string (*servproto can be NULL)
+ * @returns -1 on error, or
+ *          positive port number (in network byte order)
+ *
+ * @note aliases are not read from the file.
```
[the remark this answers is not in the document]
Comment (ahgamut) [reply libc/dns/servicestxt.c:135]: You are right, `LookupServicesByName` does need to read the aliases even if it does not store them.

**Comment by ahgamut** [comment]
@jart I have pushed commit(s) to the `servent-impl` branch after reading your comments.

> I noticed there's no unit tests for this change. Here's how that could have worked:

Thanks for providing an outline for the testing. I'll submit another PR with tests for `getservby*`, or even add them on the current `servent-impl` branch if that's okay.

>I would support your decision if you chose to make the buffer larger. It doesn't change the fact that the number is officially still 15. It's more of a Postel's Maxim kind of thing in being generous towards nonconformity. But in that case I'd argue that unlimited malloc is much better than defining another arbitrary but somewhat higher limit.

My preference for a static buffer for `struct servent.s_name` over unlimited `malloc` is because 

* the similarity with `gethostby*` code, and
* avoid the bookkeeping (like the `free`/`strdup`/`free` for `localproto`) to ensure names were being written correctly.
* I tried using `malloc` everywhere (including aliases) at first and got stuck.

Currently I'm just using `DNS_NAME_MAX` as the size limit for `struct servent.s_name` .  
I'm okay using `malloc` as well (in fact it is likely it will be needed for filling out the aliases).

File: libc/dns/servicestxt.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahgamut) [reply libc/dns/servicestxt.c:66]: I missed the obvious here: of course the internal `LookupServices*` functions can be sane, the network byte order stuff can be handled internally in `getservby*` functions.

**Pull request merged**